### ISSUE TITLE
Unpredictable reboots and high response times on Heroku

### ISSUE THREAD

**Issue by varunsrin** [Shopify/dashing#502]
I've noticed that my Heroku instance gets rebooted fairly often, losing of the data that my dashboard has, which is quite inconvenient. 

I seem to hit a lot of the following errors: 

<code>» 22:51:19.940 2015-01-14 06:51:19.719058+00:00 heroku router - - at=error code=H12 desc="Request timeout" method=GET path="/events" host=dashing.mysite.com request_id=50cc241e-2410-493c-a07b-1d4948b90480 fwd="50.168.188.129/c-50-168-188-129.hsd1.ca.comcast.net" dyno=web.1 connect=0ms service=30003ms status=503 bytes=1306 CriticalHigh Response Time Context</code>

Is there any reason events would be waiting so long to respond? Most of my jobs that fetch event data fire only once every 15 - 30 minutes, whereas I see the above errors 2-3x a minute, and they happen consistently for hours, if anything is pointing to the dashboard. 

I also seem to hit a lot of 503's in the rufus scheduler, roughly about 15-16 of these over the last day: 

<code>
» 17:43:06.864 2015-01-14 01:43:06.454631+00:00 app web.1 - - scheduler caught exception:
» 17:43:06.864 2015-01-14 01:43:06.454633+00:00 app web.1 - - 503 Service Unavailable
</code>

Usually after a couple of them, my Heroku app gets rebooted - any idea why the event stream takes so long to render? And if that's what's causing the data problem?

**Comment by tylermauthe** [comment]
Is it a Heroku free instance?

They go to sleep after 1 hour of inactivity:
https://devcenter.heroku.com/articles/dynos#dyno-sleeping

You may be able to skirt this issue for a while with:
https://coderwall.com/p/u0x3nw/avoid-heroku-idling-with-new-relic-pings

Hope this helps!

**Comment by varunsrin** [comment]
Actually, this is happening on a paid instance with two dynos. 

What I noticed that's interesting is that Heroku Metrics don't work claiming that two or more dynos are needed to run Metrics, even though the dyno page says there are two 1x dynos running and is billing us $34/month for it. 

This is my first Sinatra deployment on Heroku, is there anything specific that needs to be configured to get the dashing server to utilize both dynos?

**Comment by varunsrin** [comment]
Further investigation:

* Disabling GitHub authentication had no impact
* Installed New Relic, but all transactions in New Relic average 300 ms, with the highest being 2k ms. Nothing in New Relic indicates the cause of the H12 errors - in fact, I can't even see any request queueing that takes that long. 

But I'm still seeing frequent H12 timeout events in the heroku logs that return a 503 with a 30,000 ms wait time. This happens whenever I have any dashboard open.

I'm assuming that the request is being passed on to my server and Sinatra+Thin are not sending a response back to Heroku routers in < 30s.  

I'm a little surprised that these long running events are not showing up in New Relic - I'm not very familiar with Sinatra, but I'm going to keep looking into this....

**Comment by tylermauthe** [comment]
I'm sorry, I should have read the details of your inquiry more closely the first time. The events service streams data to the dashboard via server-sent events, so you are seeing H12 because Heroku is toasting the request.

https://devcenter.heroku.com/articles/request-timeout#long-polling-and-streaming-responses

You are probably already using the Cedar stack, as it is the default buildpack. Use it instead if you specified something else.

Otherwise, I think the issue may actually be that you aren't sending data often enough:

> Cedar supports HTTP 1.1 features such as long-polling and streaming responses. An application has an initial 30 second window to respond with a single byte back to the client. However, each byte transmitted thereafter (either received from the client or sent by your application) resets a rolling 55 second window. If no data is sent during the 55 second window, the connection will be terminated.

Add a heartbeat job like this and let me know if it resolves the H12 issues:
```
SCHEDULER.every '50s', :first_in => 0 do
 	send_event('heartbeat', { pulse: true })
end
```

**Comment by varunsrin** [comment]
So here's what I've seen so far: 

* I've been using Cedar (have seen some H15 errors occasionally, which are cedar specific) 
* Heartbeat event has not helped
* Scaling dynos from 1 -> 2 has not helped at all 
* Adding rack-timeout (15s) didn't trigger any timeout exceptions on the dynos

I think the request is getting lost somewhere between the heroku router and the dyno, because otherwise I'd expect rack-timeout to be raising exceptions. 

I don't fully understand the client-server pattern for how events work in dashing, going to dig into that more today.... perhaps there's something in there that I'm missing. 

I'm copying a few lines over from the latest logs to give you an idea of frequency 

```
 10:50:11.365 2015-01-23 18:50:10.928346+00:00 app web.1 - - 23.4.18.10/b-34-5-68-10.hdf1.ca.comcast.net - - [23/Jan/2015 18:50:10] "GET /events HTTP/1.1" 200 - 46.9910
» 10:50:14.581 2015-01-23 18:50:14.256238+00:00 app web.1 - - source=rack-timeout id=c7a15979-77a8-40ee-b2e8-8f784b1b36c8 wait=4ms timeout=15000ms state=ready at=info
» 10:51:01.180 2015-01-23 18:51:00.839525+00:00 heroku router - - at=error code=H12 desc="Request timeout" method=GET path="/events" host=foo.bar.com request_id=c7a15979-77a8-40ee-b2e8-8f784b1b36c8 fwd="23.4.18.10/b-34-5-68-10.hdf1.ca.comcast.net" dyno=web.1 connect=1ms service=46582ms status=503 bytes=1387 CriticalHigh Response Time
» 10:51:01.265 2015-01-23 18:51:00.838959+00:00 app web.1 - - 23.4.18.10/b-34-5-68-10.hdf1.ca.comcast.net - - [23/Jan/2015 18:51:00] "GET /events HTTP/1.1" 200 - 46.5814
» 10:51:04.382 2015-01-23 18:51:04.074732+00:00 app web.1 - - source=rack-timeout id=ca31ccf4-626a-4f88-bddd-318172b5baad wait=7ms timeout=15000ms state=ready at=info
» 10:51:51.235 2015-01-23 18:51:50.847724+00:00 app web.1 - - 23.4.18.10/b-34-5-68-10.hdf1.ca.comcast.net - - [23/Jan/2015 18:51:50] "GET /events HTTP/1.1" 200 - 46.7719
» 10:51:51.452 2015-01-23 18:51:50.842524+00:00 heroku router - - at=error code=H12 desc="Request timeout" method=GET path="/events" host=foo.bar.com request_id=ca31ccf4-626a-4f88-bddd-318172b5baad fwd="23.4.18.10/b-34-5-68-10.hdf1.ca.comcast.net" dyno=web.1 connect=3ms service=46769ms status=503 bytes=1387 CriticalHigh Response Time
» 10:51:54.393 2015-01-23 18:51:54.085162+00:00 app web.1 - - source=rack-timeout id=c848edc5-086c-4fab-a8f5-527a33d64ca3 wait=9ms timeout=15000ms state=ready at=info
» 10:52:41.777 2015-01-23 18:52:41.457082+00:00 app web.1 - - 23.4.18.10/b-34-5-68-10.hdf1.ca.comcast.net - - [23/Jan/2015 18:52:41] "GET /events HTTP/1.1" 200 - 47.3707
```

**Comment by itgoujie2** [comment]
i am having the exact same issue. I got H12+503 error some times, I do have set up New Relic to ping my app to make sure it is not idling. And New Relic will not catch this error. If you have found the solution, please please let me know

**Comment by tylermauthe** [comment]
Try asking Heroku support.

**Comment by syednasir** [comment]
Did that issue get resolved? We're having similar issues on heroku (with event streams) and we have  tried scaling dynos, rack-tineout, heartbeat etc.

**Comment by tylermauthe** [comment]
Not sure if the Timeout issue has been solved -- but maybe this could be solved by trying to recover gracefully when there is a timeout. Your data should be written to a `history.yml` file when Dashing is shut down gracefully, which on Heroku is deleted because of its [Ephemeral Filesystem](https://devcenter.heroku.com/articles/dynos#ephemeral-filesystem).

If PR #330 were reviewed and merged, then data could be stored Redis, allowing it to persist...

**Comment by ryanprater** [comment]
I've been experiencing the same issue for the past 6 months but haven't had time to fully address the issue. From my research I believe the issue to be gevent and multi-threading with slow clients. See the following links for context:
https://github.com/benoitc/gunicorn/issues/588
https://discussion.heroku.com/t/gunicorn-dyno-death-spiral/136
http://blog.etianen.com/blog/2014/01/19/gunicorn-heroku-django/
Will report back once I am able to investigate.

**Comment by itgoujie2** [comment]
I switched to aws ec2, never got timeout error even once, same app, Heroku needs to take this issue seriously..

**Comment by nitronick600** [comment]
I'm seeing this as well.  I've tried the heartbeat as well as New Relic without luck.  Currently, I'm instructing the customer to refresh the dashboard periodiclly, but there must be a better way.  

I've contacted Heroku support, we'll see what they say.

**Comment by ryanprater** [comment]
Hi all,

I spent 6+ months this year going back and forth with Heroku Support, and finally got a solid answer out of them. Please have a look at my SO answer here: http://stackoverflow.com/questions/29088113/heroku-sporadic-high-response-time/32465651#32465651 

Hopefully that helps. Good luck!

**Comment by tylermauthe** [comment]
@ryanprater Thanks a ton for your insights into this issue! Very helpful information and it sounds like the best theory yet that explains the symptoms described in this issue.

Good luck resolving this with Heroku! Please create a new issue if you think that Dashing could be improved to play more nicely with Heroku, but it sounds like any old Ruby (2.1+) app would suffer the same problems.
